atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java

```diff
@@ -2551,6 +2568,183 @@ public void exceptionThrownWhenTooManyPostFilterIterationsOccur() {
+        TimelockService timelockService = spy(txManager.getTimelockService());
+        Transaction txn = getSnapshotTransactionWith(
+                timelockService, ImmutableMap.of(TABLE, Optional.of(ConflictHandler.SERIALIZABLE)));
+        txn.put(TABLE, ImmutableMap.of(TEST_CELL, TEST_VALUE));
+        verifyLockWasCalledWithLocksAndMetadataWhenCommitting(
+                txn,
+                timelockService,
+                LocksAndMetadata.of(
+                        ImmutableSet.of(AtlasRowLockDescriptor.of(TABLE.getQualifiedName(), TEST_CELL.getRowName())),
+                        Optional.empty()));
+    }
+
+    @Test
+    public void lockCollectionWithoutMetadataIsCorrectForRandomWrites() {
+        long randomSeed = System.currentTimeMillis();
+        int numTables = 10;
+        int numCellsPerTable = 100;
+        int totalWrites = 200;
+
+        Random random = new Random(randomSeed);
+
+        List<ConflictHandler> availableConflictHandlers = Arrays.asList(ConflictHandler.values());
+        List<TableReference> tables = IntStream.range(0, numTables)
+                .mapToObj(i -> {
+                    TableReference tableRef = TableReference.create(Namespace.create("random_test"), "table" + i);
+                    keyValueService.createTable(tableRef, AtlasDbConstants.GENERIC_TABLE_METADATA);
+                    return tableRef;
+                })
+                .collect(Collectors.toUnmodifiableList());
+        Map<TableReference, Optional<ConflictHandler>> tableToConflictHandler = KeyedStream.of(tables.stream())
+                .map(tableRef ->
+                        Optional.of(availableConflictHandlers.get(random.nextInt(availableConflictHandlers.size()))))
+                .collectToMap();
+        // Ensure that some rows have more than one cell
+        List<Cell> cells = IntStream.range(0, numCellsPerTable)
+                .mapToObj(i -> Cell.create(PtBytes.toBytes("row" + (i % 3)), PtBytes.toBytes("column" + i)))
+                .collect(Collectors.toUnmodifiableList());
+
+        TimelockService timelockService = spy(txManager.getTimelockService());
+        Transaction txn = getSnapshotTransactionWith(timelockService, tableToConflictHandler);
+
+        Map<TableReference, Set<Cell>> cellsWithWrites = new HashMap<>();
+        IntStream.range(0, totalWrites).forEach(_unused -> {
+            TableReference table = tables.get(random.nextInt(tables.size()));
+            Cell cell = cells.get(random.nextInt(cells.size()));
+            if (random.nextBoolean()) {
+                txn.put(table, ImmutableMap.of(cell, TEST_VALUE));
+            } else {
+                txn.delete(table, ImmutableSet.of(cell));
+            }
+            cellsWithWrites.computeIfAbsent(table, _ignored -> new HashSet<>()).add(cell);
+        });
+
+        Set<LockDescriptor> expectedLocks = getExpectedLocks(tableToConflictHandler, cellsWithWrites);
+        verifyLockWasCalledWithLocksAndMetadataWhenCommitting(
```

Comment:
How much harder would it be for us to add metadata occasionally?

Reply:
if it's too hard to due to having row locks, then let's forget it, but would be nice to have just for additional coverage.

Reply:
Yeah the original random test had metadata but it would nearly always result in commit failure due to duplicate metadata within a row. I think the property based tests cover metadata sufficiently. The random tests is more to verify that the regular lock collection works as it did previously.

Reply:
I'll sprinkle in some metadata